<a href="https://colab.research.google.com/github/Dekromatia/RSA/blob/main/hw3task4%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [35]:
import numpy as np
import pandas as pd
__author__ = 'Chris Potts'


class RSA:
    """Implementation of the core Rational Speech Acts model.

    Parameters
    ----------
    lexicon : `np.array` or `pd.DataFrame`
        Messages along the rows, states along the columns.
    prior : array-like
        Same length as the number of colums in `lexicon`.
    costs : array-like
        Same length as the number of rows in `lexicon`.
    alpha : float
        The temperature parameter. Default: 1.0
    """
    def __init__(self, lexicon, prior, costs, alpha=1.0):
        self.lexicon = lexicon
        self.prior = np.array(prior)
        self.costs = np.array(costs)
        self.alpha = alpha

    def literal_listener(self):
        """Literal listener predictions, which corresponds intuitively
        to truth conditions with priors.

        Returns
        -------
        np.array or pd.DataFrame, depending on `self.lexicon`.
        The rows correspond to messages, the columns to states.

        """
        return rownorm(self.lexicon * self.prior)

    def speaker(self):
        """Returns a matrix of pragmatic speaker predictions.

        Returns
        -------
        np.array or pd.DataFrame, depending on `self.lexicon`.
        The rows correspond to states, the columns to states.
        """
        lit = self.literal_listener().T
        utilities = self.alpha * (safelog(lit) + self.costs)
        return rownorm(np.exp(utilities))

    def listener(self):
        """Returns a matrix of pragmatic listener predictions.

        Returns
        -------
        np.array or pd.DataFrame, depending on `self.lexicon`.
        The rows correspond to messages, the columns to states.
        """
        spk = self.speaker().T
        return rownorm(spk * self.prior)


def rownorm(mat):
    """Row normalization of np.array or pd.DataFrame"""
    return (mat.T / mat.sum(axis=1)).T


def safelog(vals):
    """Silence distracting warnings about log(0)."""
    with np.errstate(divide='ignore'):
        return np.log(vals)


if __name__ == '__main__':
    """Examples"""

    from IPython.display import display


    def display_reference_game(mod):
        d = mod.lexicon.copy()
        d['costs'] = mod.costs
        d.loc['prior'] = list(mod.prior) + [""]
        d.loc['alpha'] = [mod.alpha] + [" "] * mod.lexicon.shape[1]
        display(d)


    # TASK 4.1:
    msgs = ['streped', 'checkered' , 'red', 'with_pillows']
    states = ['r1', 'r2', 'r3', 'r4']
    lex = pd.DataFrame([
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 1.0],
        [1.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0]],
         index=msgs, columns=states)
    print('TASK 4.1 ')
    print("="*70 + "\nEven priors and all-0 message costs\n")
    basic_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[0.0, 0.0, 0.0, 0.0])

    display_reference_game(basic_mod)

    print("\nLiteral listener")
    display(basic_mod.literal_listener())

    print("\nPragmatic speaker")
    display(basic_mod.speaker())

    print("\nPragmatic listener")
    display(basic_mod.listener())

    # TASK 4.2 changed prior1:
    msgs = ['streped', 'checkered' , 'red', 'with_pillows']
    states = ['r1', 'r2', 'r3', 'r4']
    lex = pd.DataFrame([
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 1.0],
        [1.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0]],
         index=msgs, columns=states)
    print('TASK 4.2 changed prior1')
    # print("="*70 + "\nEven priors and all-0 message costs\n")
    basic_mod = RSA(lexicon=lex, prior=[0.45, 0.05, 0.05, 0.45], costs=[0.0, 0.0, 0.0, 0.0])

    display_reference_game(basic_mod)

    print("\nLiteral listener")
    display(basic_mod.literal_listener())

    print("\nPragmatic speaker")
    display(basic_mod.speaker())

    print("\nPragmatic listener")
    display(basic_mod.listener())

# TASK 4.2 changed prior2:
    msgs = ['streped', 'checkered' , 'red', 'with_pillows']
    states = ['r1', 'r2', 'r3', 'r4']
    lex = pd.DataFrame([
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 1.0],
        [1.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0]],
         index=msgs, columns=states)
    print('TASK 4.2 changed prior2')
    # print("="*70 + "\nEven priors and all-0 message costs\n")
    basic_mod = RSA(lexicon=lex, prior=[0.05, 0.45, 0.45, 0.05], costs=[0.0, 0.0, 0.0, 0.0])

    display_reference_game(basic_mod)

    print("\nLiteral listener")
    display(basic_mod.literal_listener())

    print("\nPragmatic speaker")
    display(basic_mod.speaker())

    print("\nPragmatic listener")
    display(basic_mod.listener())

    # TASK 4.2 changed costs+:
    msgs = ['streped', 'checkered' , 'red', 'with_pillows']
    states = ['r1', 'r2', 'r3', 'r4']
    lex = pd.DataFrame([
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 1.0],
        [1.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0]],
         index=msgs, columns=states)
    print('TASK 4.2 changed costs+')
    # print("="*70 + "\nEven priors and all-0 message costs\n")
    basic_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[0.0, 10, 10, 0.0])

    display_reference_game(basic_mod)

    print("\nLiteral listener")
    display(basic_mod.literal_listener())

    print("\nPragmatic speaker")
    display(basic_mod.speaker())

    print("\nPragmatic listener")
    display(basic_mod.listener())

     # TASK 4.2 changed costs-:
    msgs = ['streped', 'checkered' , 'red', 'with_pillows']
    states = ['r1', 'r2', 'r3', 'r4']
    lex = pd.DataFrame([
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 0.0, 1.0, 1.0],
        [1.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0]],
         index=msgs, columns=states)
    print('TASK 4.2 changed costs-')
    # print("="*70 + "\nEven priors and all-0 message costs\n")
    basic_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[-1, 0.0, 0.0, -1])

    display_reference_game(basic_mod)

    print("\nLiteral listener")
    display(basic_mod.literal_listener())

    print("\nPragmatic speaker")
    display(basic_mod.speaker())

    print("\nPragmatic listener")
    display(basic_mod.listener())


TASK 4.1 
Even priors and all-0 message costs



r1    r2    r3    r4 costs
streped       1.00     0     0     0     0
checkered     0.00     0     1     1     0
red           1.00     1     0     0     0
with_pillows  0.00     0     0     1     0
prior         0.25  0.25  0.25  0.25      
alpha         1.00


Literal listener


r1   r2   r3   r4
streped       1.0  0.0  0.0  0.0
checkered     0.0  0.0  0.5  0.5
red           0.5  0.5  0.0  0.0
with_pillows  0.0  0.0  0.0  1.0


Pragmatic speaker


streped  checkered       red  with_pillows
r1  0.666667   0.000000  0.333333      0.000000
r2  0.000000   0.000000  1.000000      0.000000
r3  0.000000   1.000000  0.000000      0.000000
r4  0.000000   0.333333  0.000000      0.666667


Pragmatic listener


r1    r2    r3    r4
streped       1.00  0.00  0.00  0.00
checkered     0.00  0.00  0.75  0.25
red           0.25  0.75  0.00  0.00
with_pillows  0.00  0.00  0.00  1.00

TASK 4.2 changed prior1


r1    r2    r3    r4 costs
streped       1.00     0     0     0     0
checkered     0.00     0     1     1     0
red           1.00     1     0     0     0
with_pillows  0.00     0     0     1     0
prior         0.45  0.05  0.05  0.45      
alpha         1.00


Literal listener


r1   r2   r3   r4
streped       1.0  0.0  0.0  0.0
checkered     0.0  0.0  0.1  0.9
red           0.9  0.1  0.0  0.0
with_pillows  0.0  0.0  0.0  1.0


Pragmatic speaker


streped  checkered       red  with_pillows
r1  0.526316   0.000000  0.473684      0.000000
r2  0.000000   0.000000  1.000000      0.000000
r3  0.000000   1.000000  0.000000      0.000000
r4  0.000000   0.473684  0.000000      0.526316


Pragmatic listener


r1    r2    r3    r4
streped       1.00  0.00  0.00  0.00
checkered     0.00  0.00  0.19  0.81
red           0.81  0.19  0.00  0.00
with_pillows  0.00  0.00  0.00  1.00

TASK 4.2 changed prior2


r1    r2    r3    r4 costs
streped       1.00     0     0     0     0
checkered     0.00     0     1     1     0
red           1.00     1     0     0     0
with_pillows  0.00     0     0     1     0
prior         0.05  0.45  0.45  0.05      
alpha         1.00


Literal listener


r1   r2   r3   r4
streped       1.0  0.0  0.0  0.0
checkered     0.0  0.0  0.9  0.1
red           0.1  0.9  0.0  0.0
with_pillows  0.0  0.0  0.0  1.0


Pragmatic speaker


streped  checkered       red  with_pillows
r1  0.909091   0.000000  0.090909      0.000000
r2  0.000000   0.000000  1.000000      0.000000
r3  0.000000   1.000000  0.000000      0.000000
r4  0.000000   0.090909  0.000000      0.909091


Pragmatic listener


r1    r2    r3    r4
streped       1.00  0.00  0.00  0.00
checkered     0.00  0.00  0.99  0.01
red           0.01  0.99  0.00  0.00
with_pillows  0.00  0.00  0.00  1.00

TASK 4.2 changed costs+


r1    r2    r3    r4 costs
streped       1.00     0     0     0     0
checkered     0.00     0     1     1    10
red           1.00     1     0     0    10
with_pillows  0.00     0     0     1     0
prior         0.25  0.25  0.25  0.25      
alpha         1.00


Literal listener


r1   r2   r3   r4
streped       1.0  0.0  0.0  0.0
checkered     0.0  0.0  0.5  0.5
red           0.5  0.5  0.0  0.0
with_pillows  0.0  0.0  0.0  1.0


Pragmatic speaker


streped  checkered       red  with_pillows
r1  0.000091   0.000000  0.999909      0.000000
r2  0.000000   0.000000  1.000000      0.000000
r3  0.000000   1.000000  0.000000      0.000000
r4  0.000000   0.999909  0.000000      0.000091


Pragmatic listener


r1        r2        r3        r4
streped       1.000000  0.000000  0.000000  0.000000
checkered     0.000000  0.000000  0.500023  0.499977
red           0.499977  0.500023  0.000000  0.000000
with_pillows  0.000000  0.000000  0.000000  1.000000

TASK 4.2 changed costs-


r1    r2    r3    r4 costs
streped       1.00     0     0     0    -1
checkered     0.00     0     1     1     0
red           1.00     1     0     0     0
with_pillows  0.00     0     0     1    -1
prior         0.25  0.25  0.25  0.25      
alpha         1.00


Literal listener


r1   r2   r3   r4
streped       1.0  0.0  0.0  0.0
checkered     0.0  0.0  0.5  0.5
red           0.5  0.5  0.0  0.0
with_pillows  0.0  0.0  0.0  1.0


Pragmatic speaker


streped  checkered       red  with_pillows
r1  0.423883   0.000000  0.576117      0.000000
r2  0.000000   0.000000  1.000000      0.000000
r3  0.000000   1.000000  0.000000      0.000000
r4  0.000000   0.576117  0.000000      0.423883


Pragmatic listener


r1        r2        r3        r4
streped       1.000000  0.000000  0.000000  0.000000
checkered     0.000000  0.000000  0.634471  0.365529
red           0.365529  0.634471  0.000000  0.000000
with_pillows  0.000000  0.000000  0.000000  1.000000

4.1 Игра:
Вы дизайнер и выбрали для килиента  несколько диванов: красный диван в полоску, красный диван, клетчатый диван и клетчатый диван с подушками.
Вы отпрвляете фото диванов клиенту, с какой вероятностью при каждом из высказываний m клиетнт выберет ту или иную отправленную ему картинку. 
состояния:
r1 = red streped sofa,
r2 = red sofa, 
r3 = checkered sofa,
r4 = checkered sofa with pillows

высказывания m1 = streped, 
m2 = checkerd, 
m3= red, 
m4=with pillows, 
priot = 0,  alpha = 1, cost = 0 

 После вычислений с помощью RSA у нас получается 2 импликатуры:
 1)если мы говорим "красный"  клиент посмотрит на 2 фоттки r1 и r2 с вероятностью 0.75 выберет r2, и лишь с вероятностью 0.25 подумает о r1, т.к. для описания r1  мы бы как прогматичный говорящий выбрали высказывание m1 - "полосатый"
2)если мы говорим "клетчатый" то клиент посмотрит на 2 фото r3 и r4, и с вероятностью 0.75 выберет r3, вероятность выбора r4 составит 0.25 т.к. для описания r4 у нас есть высказывание "с подушками"

4.2
1)с увеличением исходной вероятности состония r вероятность выбора этого состояния в случае с импликатурой тоже увеличивается, и наоборот с уменьшением prior для r вероятность выбоа этого r при высказывании m уменьшается

если мы изменим prior для r2 и r3 на 0.05 (если заменить на 0.1 вероятность ыерного выбоа будет 0.39 а не верного 0.61, поэтому я выбрала более яркий пример с 0.05 ) - импликатуры иссчезнут т.к. слушатель будет выбирать r1  не зависимо от того что мы скажем "полосатый"(вероятность выбора - 1.0) или "красный"(0.81)
Во втором случае слушатель будет выбирать r4 не зависимо от того что мы скажем 
"клетчатый"(0.81) или "с подушками"(1.0)
2)если мы изменим prior для r1 и r4 на 0.05 то импликатура тоже исчезнет т.к. каждое состояние m будет соответствовать определеннму высказыванию r 
r1-полосатый, 
r2-красный, 
r3-в клетку, 
r4-с подушками
Следуя такой же логике можно убрать 1 из 2 импликатур 

2)c увеличением цены высказывания вырастает вероятность выбора этого высказывания для наименее подходящего состояния r согласно максиме количества и уменьшается для наиболее подходящего, однако это происходит до тех пор пока вероятности не станут одинаковыми т.е. 0.5 , далее можео повышать стоимость на сколько угодно но ничего не изменится.
(например и при стоимости 14 и при стоимости 100 для "клетчатого" вероятность выбора r3 или r4  и для "красного" вероятность выбора r1 или r2 будет=0.5)
до достижения стоимости = 14 
при высказывании "клетчатый" вероятность выбора состояния r3 будет постепенно падать(до 0.5) с увеличением цены этого высказывания, а вероятность выбора состояния r4 - рости(до 0.5) с ростом цены.
аналогично для высказывания "красный" вероятность выбора состояния r2 будет постепенно падать, а вероятность выбора r1 рости,  с увеличением цены высказывания.
тот такой же эфект для этих же высказыванйи мы сожем наблюдать если  изменим цену высказываний "в полоску" и "с подушками" на отрицательное значение например -1
т.е. с уменьшением цены высказывания вероятность выбора этого высказывания уменьшается даже в случае если оно является единственно верным, как например в случае высказывания "клетчатый" для r3

Вывод: можено сказать что при -14 для высказываний "в полоску" и "с подушками" и +14 "клетчатый" и "крвсный" импликатруры исчезают абсолютно и выбор прагматичного слушателя становится абсолютоно рандомным, как если бы мы подбросили манетку.


